### PR TITLE
[Feature-11440][Flink Stream] Use flink SDK replace the shell cli

### LINKED ISSUE #11440: [Feature][Flink Stream] Use flink SDK replace the shell cli
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

I want to participate in it.
relate to #11352 

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by shangeyao**
close #11440

### CODE REVIEW

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/pom.xml [hunk 1]
```diff
@@ -15,19 +15,27 @@
...
     <artifactId>dolphinscheduler-task-flink-stream</artifactId>
     <packaging>jar</packaging>
 
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <flink.version>1.15.1</flink.version>
+        <kerb.version>2.0.1</kerb.version>
+        <log4j.version>2.18.0</log4j.version>
+        <sl4j.version>1.6.1</sl4j.version>
+        <flink.shaded.hadoop.version>2.8.3-7.0</flink.shaded.hadoop.version>
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/pom.xml]: Don't manage the versions here, move them to the bom

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/CheckpointInfo.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.dolphinscheduler.plugin.task.flink.entity;
+
+public class CheckpointInfo {
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/CheckpointInfo.java:29]: Clean up this class with lombok annotations

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ParamsInfo.java [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+ */
+
+package org.apache.dolphinscheduler.plugin.task.flink.entity;
+
+import org.apache.dolphinscheduler.plugin.task.flink.enums.FlinkStreamDeployMode;
+
+import java.util.Arrays;
+import java.util.Properties;
+
+public class ParamsInfo {
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ParamsInfo.java]: Clean up this class with lombok annotations

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ResultInfo.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    private String jobId;
+
+    public ResultInfo(String appId, String jobId) {
+        this.appId = appId;
+        this.jobId = jobId;
+    }
+
+    public ResultInfo() {
+
+    }
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ResultInfo.java]: Clean up the CTORs with lombok annotations

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkStreamTask.java [hunk 5]
```diff
@@ -55,37 +67,110 @@ public void init() {
...
+        String flinkHome = System.getenv(FLINK_HOME);
+
+        if (flinkHome.isEmpty()) {
+            logger.error("Please make sure to set the FLINK_HOME environment variable.");
+        }
+
+        String hadoopConfDir = System.getenv(HADOOP_CONF_DIR);
+
+        if (hadoopConfDir.isEmpty()) {
+            logger.error("Please make sure to set the HADOOP_CONF_DIR environment variable.");
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkStreamTask.java:119]: Should throw taskException here?

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/client/IClusterClient.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+package org.apache.dolphinscheduler.plugin.task.flink.client;
+
+import org.apache.dolphinscheduler.plugin.task.flink.entity.CheckpointInfo;
+import org.apache.dolphinscheduler.plugin.task.flink.entity.ParamsInfo;
+import org.apache.dolphinscheduler.plugin.task.flink.entity.ResultInfo;
+import org.apache.dolphinscheduler.plugin.task.flink.enums.YarnTaskStatus;
+
+import java.util.List;
+public interface IClusterClient {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/client/IClusterClient.java:30]: Please add some comment for it.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+            Path applicationDir =
+                    new Path(
+                            checkNotNull(fs.getHomeDirectory()),
+                            ".flink/" + checkNotNull(applicationId) + '/');
+            if (!fs.delete(applicationDir, true)) {
+                logger.error(
+                        "Deleting yarn application files under {} was unsuccessful.",
+                        applicationDir);
+            } else {
+                logger.error(
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java]:
```suggestion
                logger.info(
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ParamsInfo.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import lombok.NoArgsConstructor;
+import org.apache.dolphinscheduler.plugin.task.flink.enums.FlinkStreamDeployMode;
+
+import java.util.Properties;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+@Data
+public class ParamsInfo {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ParamsInfo.java]:
```suggestion
public class FlinkParamsInfo {
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ResultInfo.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+@Data
+public class ResultInfo {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ResultInfo.java]:
```suggestion
public class FlinkResultInfo {
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+                        clusterClient.triggerSavepoint(flinkJobId, null, SavepointFormatType.DEFAULT);
+                Object result = savepointFuture.get(2, TimeUnit.MINUTES);
+                logger.info("flink job savepoint path: {}", result.toString());
+            } catch (Exception e) {
+                logger.error("flink job savepoint error", e);
+            }
+
+        } catch (Exception e) {
+            logger.error("flink job savepoint failed, appId:{}, jobId:{}, exception:{}", appId, jobId, e);
+            return new ResultInfo(appId, jobId);
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java]:
```suggestion
            throw new TaskException("flink job savepoint failed", e)
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+            ClusterClientProvider<ApplicationId> retrieve =
+                    clusterDescriptor.retrieve(applicationId);
+            try (ClusterClient<ApplicationId> clusterClient = retrieve.getClusterClient()) {
+                CompletableFuture<String> savepointFuture =
+                        clusterClient.triggerSavepoint(flinkJobId, null, SavepointFormatType.DEFAULT);
+                Object result = savepointFuture.get(2, TimeUnit.MINUTES);
+                logger.info("flink job savepoint path: {}", result.toString());
+            } catch (Exception e) {
+                logger.error("flink job savepoint error", e);
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java:204]: throw exception

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/factory/YarnClusterDescriptorFactory.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+                                YarnClientYarnClusterInformationRetriever.create(yarnClient),
+                                false);
+                return clusterDescriptor;
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            throw new RuntimeException("yarn mode must set param of 'hadoopConfDir'");
+        }
+    }
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/factory/YarnClusterDescriptorFactory.java]:
```suggestion
        if (StringUtils.isBlank(hadoopConfDir)) {
            throw new RuntimeException("yarn mode must set param of 'hadoopConfDir'");
        }
          try {
              YarnConfiguration yarnConf = parseYarnConfFromConfDir(hadoopConfDir);
              YarnClient yarnClient = createYarnClientFromYarnConf(yarnConf);

              YarnClusterDescriptor clusterDescriptor =
                      new YarnClusterDescriptor(
                              flinkConfig,
                              yarnConf,
                              yarnClient,
                              YarnClientYarnClusterInformationRetriever.create(yarnClient),
                              false);
              return clusterDescriptor;
          } catch (Exception e) {
              throw new RuntimeException(e);
          }
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/test/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkArgsUtilsTest.java [hunk 13]
```diff
@@ -17,116 +17,86 @@
...
 import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import java.util.List;
+import java.util.Properties;
 
 public class FlinkArgsUtilsTest {
 
-    private String joinStringListWithSpace(List<String> stringList) {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/test/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkArgsUtilsTest.java:29]: Maybe we should keep the old test for flink task, and add the new one for flink-stream task, WDYT?

Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/test/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkArgsUtilsTest.java:29]: But I have changed some parameters, and the original UT is outdated. @caishunfeng

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkStreamTask.java (hunk 5 above)
Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/FlinkStreamTask.java:119]: yeah, I correct it.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java (hunk 7 above)
Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java]: yeah, I correct it.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ParamsInfo.java (hunk 8 above)
Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ParamsInfo.java]: yeah, I correct it.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ResultInfo.java (hunk 9 above)
Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/entity/ResultInfo.java]: yeah, I correct it.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java (hunk 10 above)
Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java]: yeah, I correct it.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java (hunk 11 above)
Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/executor/AbstractClusterExecutor.java:204]: yeah, I correct it.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/factory/YarnClusterDescriptorFactory.java (hunk 12 above)
Comment (shangeyao) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-flink-stream/src/main/java/org/apache/dolphinscheduler/plugin/task/flink/factory/YarnClusterDescriptorFactory.java]: yeah, I correct it.

**Comment by EricGao888** [comment]
Could u plz fix the commit history of this PR? Thanks.

**Pull request closed without merging**